loop_workday_upload.py

```diff
@@ -0,0 +1,499 @@
+    """
+    results = []
+
+    if isinstance(widgets, Mapping) and key in widgets.keys() and widgets[key] == value:
+        results.append(widgets)
+    elif isinstance(widgets, Mapping):
+        for item in widgets:
+            results.extend(search_for_key_value_pair(widgets[item], key, value))
+    elif isinstance(widgets, list):
+        for item in widgets:
+            results.extend(search_for_key_value_pair(item, key, value))
+
+    return results
+
+
+def sync_expense_report_line(  # pylint: disable=too-many-arguments
+    cookies: Dict[str, str], get_line_url: str, instance_id: str, line_id: str, server: str, token: str
+) -> None:
+    """
+    Sync a single expense report line from Loop to Workday
+    """
+    print("Retrieving expense report line " + line_id + " for expense report " + instance_id + " from Workday")
+    workday_response = post(
+        url="https://wd5.myworkday.com" + get_line_url + ".htmld", cookies=cookies, data={"id": line_id}, timeout=(5, 5)
+    )
+
+    if workday_response.status_code != 200:
+        print(workday_response.status_code)
+        print(workday_response.text)
+        raise ValueError("Unexpected response code from Workday")
+
+    print("Uploading expense report line " + line_id + " for expense report " + instance_id + " to Loop")
+
+    loop_response = put(
+        url=server + "/api/v1/workday/expense-reports/" + instance_id + "/lines/" + line_id,
+        json=workday_response.json(),
+        headers={
+            "Authorization": "Bearer " + token,
+            "Accept": "application/json",
+        },
+        timeout=(5, 5),
+    )
+
+    if loop_response.status_code != 200:
+        print(loop_response.status_code)
+        print(loop_response.text)
+        raise ValueError("Unexpected response code from Loop")
+
+    print(loop_response.status_code)
+    print(loop_response.text)
+
+    for attachment in loop_response.json()["attachments"]:
+        print("Downloading attachment " + attachment + " from Workday")
+        values = search_for_key_value_pair(workday_response.json(), "instanceId", "1074$" + attachment)
+
+        if len(values) != 1:
+            print(dumps(workday_response.json()))
+            print(dumps(values))
+            raise ValueError("Did not find exactly one widget")
+
+        workday_attachment_response = get(
+            url="https://wd5.myworkday.com/gatech/attachment/1074$" + attachment + "/" + values[0]["target"] + ".htmld",
+            cookies=cookies,
+            timeout=(5, 5),
+        )
+
+        if workday_attachment_response.status_code != 200:
+            print(workday_attachment_response.status_code)
+            print(workday_attachment_response.text)
+            raise ValueError("Unexpected response code from Workday")
+
+        print("Uploading attachment " + attachment + " to Loop")
+
+        loop_attachment_response = post(
+            url=server + "/api/v1/workday/attachments/" + attachment,
```

Comment:
nit: I prefer f-strings for concatenation like this

Reply:
Good call, updated everywhere